design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+* remove functionalities or specific parts of the API
+* provide non-stable or technology preview features
+* the different APIs can evolve separately
+
+#### Extensions
+
+To better acommodate and separate the core API from extensions built on top of the API, developers should consider the following path format for extensions:
+
+```
+/apis/{API_GROUP}/extensions/{version}
+```
+
+This separation aims to avoid confusion when consuming stable and official Keycloak APIs so that clients can differentiate if they are using some API provided by a third-party with no support from Keycloak team.
+
+#### Technology Preview
+
+To better acommodate and separate the core API from technology preview features, developers should use a beta version as the API version:
+
+```
+/apis/{API_GROUP}/v1beta1
+```
+
+Technology preview APIs are related with features that have great potential to become fully supported. They provide the latest work in some area of interest and are offered with the intent to gather feedback from the community about new functionalities.
+
+#### Store Resources
+
+A resource store provides operations for a given resource. For example:
+
+```
+/apis/accounts/v1/1234/applications
+```
+
+The example above is representing a resource store path from where we can manage the `applications` for a user account with ID `1234`. Based on this path, clients can perform different operations based on the HTTP verbs or any controller resource defined to the store. Usually this type of resource is associated with CRUD operations for a specific resource.
+
+When a resource fits under this category, the name should be a plural noun.
+
+#### Controller Resources
+
+Under some circumstances, the HTTP verbs are not enough to represent an operation that can be performed in a resource. In this case,
+the resource path may indicate a specific action to be taken when accessed.
+
+```
+/apis/accounts/v1/1234/disable
+```
+
+The example above shows the usage of `disable` to operate accounts in order to disable them.
+
+When a resource fits under this category, the name should be a verb, using camel-case if necessary.
+
+### Resource Representation
+
+The standard format for resource representation is JSON. There is no rule for schemas or the semantics used by a resource representation, as long as they follow the JSON syntax properly.
+
+However, it is important to taken into account the following principles when defining resource representations:
+
+* Keep it simple for clients, so that a representation can be easily built and processed when making requests/responses to the API
+* Be aware of extensive representations as they can affect overall performance and usability, if possible review your API to see if resources can be decomposed into smaller ones
+* Avoid deeply nested JSON objects
+* Avoid using generic resource representations
+
+### Http Status Codes
+
+The tables below defines the HTTP status code that clients should expect from APIs.
```

Comment:
In each table, should we also state what headers should be included in the response. It's not applicable for each of the stated codes but e.g. `405` response should contain `Allow` header, `307` code should have `Location` header, etc.

Reply:
I will leave that open of others comment and see how far we want to go. IMO, some of these codes have implicit headers.

Reply:
@vmuzikar I think it's too much. As Pedro mentioned, some of these codes are implicit. Let's take as an example HTTP 405. If we look at RFC 7231 (https://tools.ietf.org/html/rfc7231#section-6.5.5) we will find out that `Allow` will be present in the header.

Reply:
@abstractj Ok, I agree - let's not overcomplicate this. You can mark this as resolved (I don't seem to have the permissions).